### ISSUE TITLE
Very slow evaluation of imported InceptionV3 from Keras+Theano

### ISSUE THREAD

**Issue by rohrl** [deeplearning4j/deeplearning4j#2832]
Hi,
I followed your instructions to make a local build of 0.7.3-SNAPSHOT on my Windows machine (CPU-accelerated).
Then imported Inception V3 from Keras+Theano.
I use following code to classify images but the network.output() method takes 5+ seconds to evaluate (I tried many different images).

```
        BufferedImage thumbnail = imagePage.getImageThumbnail(IMG_HEIGHT);
        // load image as nd4j matrix
        NativeImageLoader imageLoader = new NativeImageLoader(IMG_HEIGHT, IMG_WIDTH, 3);
        INDArray input = imageLoader.asMatrix(thumbnail);
        // scale scalars to range (-1, 1)
        ImagePreProcessingScaler scaler = new ImagePreProcessingScaler(-1, 1);
        scaler.preProcess(input);
        // evaluate network outputs for the input matrix
        Stopwatch sw = Stopwatch.createStarted();
        INDArray output = network.output(false, input)[0];
        log.info("Evaluation took " + sw);
```

I had a look at heap in visualvm but it was only using at most 50% of Xmx so it's not a memory problem.
CPU usage was 100%.

**Comment by rohrl** [comment]
Also, could you provide some reference point - what would you expect the evaluation time to be based on your experience?

**Comment by raver119** [comment]
Can you please provide model configuration (keras is ok) and link to Stopwatch implementation.

**Comment by raver119** [comment]
Also, can you please write a loop that'll have network.output() call within, and give us profiler screenshots?

**Comment by rohrl** [comment]
I ran the following code in a loop iterating over images in a folder (single class with main method, no extra stack).
```
        Path folder = Paths.get(dataFolderName);

        System.out.println("Loading network...");
        ComputationGraph network = ModelSerializer.restoreComputationGraph(Paths.get(networkFileName).toFile(), false);
        System.out.println("... network loaded.");

        Files.list(folder).forEach(silenceConsumer(f ->
                                                   {
                                                       BufferedImage image = ImageIO.read(f.toFile());

                                                       // load image as nd4j matrix
                                                       NativeImageLoader imageLoader = new NativeImageLoader(299, 299, 3);
                                                       INDArray input = imageLoader.asMatrix(image);

                                                       // scale scalars to range (-1, 1)
                                                       ImagePreProcessingScaler scaler = new ImagePreProcessingScaler(-1, 1);
                                                       scaler.preProcess(input);

                                                       // evaluate network outputs for the input matrix
                                                       Stopwatch sw = Stopwatch.createStarted();
                                                       INDArray output = network.output(false, input)[0];
                                                       System.out.println("Evaluation took " + sw);

                                                   }));
```
Here is the profiler screen:
![dl4j_profiler](https://cloud.githubusercontent.com/assets/14173005/22770289/71ceebfa-eee3-11e6-924e-e02b75280ccb.png)
Here's the stdout:
```
C:\dev-env\Tools\Java\jdk1.8.0_92\bin\java -Xmx2g -Xms2g -Didea.launcher.port=7534 "-Didea.launcher.bin.path=C:\Program Files (x86)\JetBrains\IntelliJ IDEA 2016.3.4\bin" -classpath C:\Users\rohrl\AppData\Local\Temp\classpath1.jar -Dfile.encoding=US-ASCII com.intellij.rt.execution.application.AppMain com.nuix.load.image.NetEvalIsolated
Loading network...
log4j:WARN No appenders could be found for logger (org.nd4j.linalg.factory.Nd4jBackend).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
... network loaded.
Evaluation took 1.326 s
Evaluation took 945.6 ms
Evaluation took 860.6 ms
Evaluation took 823.3 ms
Evaluation took 1.096 s
Evaluation took 946.7 ms
Evaluation took 881.9 ms
Evaluation took 1.135 s
Evaluation took 845.1 ms
Evaluation took 793.2 ms
Evaluation took 832.0 ms
Evaluation took 1.137 s
Evaluation took 822.1 ms
Profiler Agent: Waiting for connection on port 5140 (Protocol version: 15)
Profiler Agent: Established connection with the tool
Evaluation took 969.6 ms
Profiler Agent: Local accelerated session
Evaluation took 1.123 s
*** Profiler engine warning: Failed to lookup cached class org.trypticon.resourcebundlefix.ResourceBundleControlProviderImpl$ResourceBundleControlImpl$$Lambda$7/965586344
Evaluation took 1.633 s
Evaluation took 1.339 s
Evaluation took 1.177 s
Evaluation took 1.756 s
Evaluation took 1.362 s
Evaluation took 826.6 ms
Evaluation took 849.1 ms
Evaluation took 1.043 s
Evaluation took 819.2 ms
Evaluation took 784.1 ms
Evaluation took 812.5 ms
Evaluation took 1.048 s
Evaluation took 807.2 ms
Evaluation took 801.4 ms
Evaluation took 1.007 s
Evaluation took 824.9 ms
Evaluation took 781.9 ms
Evaluation took 801.6 ms
Evaluation took 1.138 s
Evaluation took 888.7 ms
Evaluation took 789.9 ms
Evaluation took 778.1 ms
Evaluation took 1.040 s
Evaluation took 796.7 ms
Evaluation took 877.5 ms
Evaluation took 836.7 ms
Evaluation took 1.050 s
Evaluation took 807.1 ms
Evaluation took 788.5 ms
Evaluation took 1.046 s
Evaluation took 904.7 ms
Evaluation took 826.0 ms
Evaluation took 787.2 ms
Evaluation took 1.045 s
Evaluation took 807.3 ms
Evaluation took 858.5 ms
Evaluation took 829.0 ms
```

**Comment by rohrl** [comment]
Some time ago I used Cafe's GoogleNet and it was taking 20 ms (50x faster) to evaluate on average, so I'd expect similar performance from dl4j

**Comment by rohrl** [comment]
FWIW here are some info logs from nd4j:
```
2017-02-09 16:55:46,127 [WorkerThread: ff66971aeb5c46d2a7d0d7b473c09f96] 102594 INFO  org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
2017-02-09 16:55:46,221 [WorkerThread: ff66971aeb5c46d2a7d0d7b473c09f96] 102688 INFO  org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for NativeOps: 4
2017-02-09 16:55:46,750 [WorkerThread: ff66971aeb5c46d2a7d0d7b473c09f96] 103217 INFO  org.nd4j.nativeblas.Nd4jBlas - Number of threads used for BLAS: 4
2017-02-09 16:55:46,750 [WorkerThread: ff66971aeb5c46d2a7d0d7b473c09f96] 103217 INFO  org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Windows 7]
2017-02-09 16:55:46,750 [WorkerThread: ff66971aeb5c46d2a7d0d7b473c09f96] 103217 INFO  org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [4]; Memory: [5.3GB];
2017-02-09 16:55:46,750 [WorkerThread: ff66971aeb5c46d2a7d0d7b473c09f96] 103217 INFO  org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
```

**Comment by mjakobus** [comment]
I also have sometimes performance problems. Most models are really fast but some models are about 20 oder 30 times slower during training (using cuda). I still not really sure what the cause of this is, but I currently suspect the BatchNorm Layers.
@rohrl I saw, that your model also has BatchNormalization layers. So perhaps this might also be the performance problem in your case?

**Comment by rohrl** [comment]
Can batch normalisation layer impact performance if I'm not training at all?

**Comment by raver119** [comment]
@mjakobus do you have dl4j model example with such perf issues you've mentioned? That sounds like first suspect.

**Comment by rohrl** [comment]
I imported VGG19 from Keras (Theano) and it's taking 10 seconds on avg to evaluate on CPU (VGG19 does not have BatchNormalization layers afaik).

@mjakobus which models are fast? did you import them to DL4J or trained from scratch? Did you use CPU backend ?

@raver119 Do you have any benchmark results of DL4J so I could learn what performance I can expect from it in general?

**Comment by tomthetrainer** [comment]
VGG-16, however on my macbook it is between 1 and 2 seconds, never as long as 10.

**Comment by eraly** [comment]
@rohrl I did a quick comparison with prediction times on VGG16 with Keras and dl4j. 
Keras took a little more than 2s per prediction on my mac book pro and dl4j averaged about 1.6s.

Now, I am not sure what you are seeing is architecture specific to Inception. It would be worth taking a look at the numbers you are seeing with VGG16. In any case, we will look into this further. But do confirm what you see for VGG16.

**Comment by turambar** [comment]
This is very unlikely to be related to model import _per se_ but perhaps intrinsic to DL4J (maybe inefficiencies in batch norm or something).

That said,@eraly's and @tomthetrainer's reports (and I can confirm similar findings), It's looking like this may be specific to a particular hardware/configuration/ND4J backend...?

@rohrl @mjakobus @eraly @tomthetrainer can each of you provide details about hardware, which ND4J backend you're using, are you using BLAS/MKL/etc.?

**Comment by tomthetrainer** [comment]
VGG-16 works well for me. 
In keras I was using theano backend
<nd4j.backend>nd4j-native</nd4j.backend>
02:11:40.118 [main] INFO  o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
Macbook pro 
Model Name:	MacBook Pro
  Model Identifier:	MacBookPro11,2
  Processor Name:	Intel Core i7
  Processor Speed:	2.2 GHz
  Number of Processors:	1
  Total Number of Cores:	4
  L2 Cache (per Core):	256 KB
  L3 Cache:	6 MB
  Memory:	16 GB

**Comment by tomthetrainer** [comment]
I have also run on Ubuntu with no issues, an AMI, a wee small tiny AMI it was as well.

**Comment by rohrl** [comment]
My times:
VGG19 was taking 5-10 sec for me
Inception v3: 0.8 - 1 sec

Both imported from Keras with Theano backend.
Using native CPU build from master.
> Loaded [CpuBackend] backend
> Number of threads used for NativeOps: 4
> Number of threads used for BLAS: 4
> Backend used: [CPU]; OS: [Windows 7]
> Cores: [4]; Memory: [5.3GB];
> Blas vendor: [OPENBLAS]

Hardware:
Windows 7 64bit VM running on Mac Pro via VMWare Fusion (using all cores and 16gb ram)
3.7 GHz Quad-Core Intel Xeon E5
32 GB 1866 MHz DDR3 ECC

@turambar @eraly 
Can you give me rough estimate how much faster would it get if I used GPU backend?
I can't test it at the moment as I don't have NVidia GPU
Thanks

**Comment by turambar** [comment]
@rohrl I think it's not unreasonable to suggest you could get at least 10X speedup and more likely even more.

**Comment by raver119** [comment]
I've found issue affecting performance for CNN-based model. Fix incoming.

**Comment by raver119** [comment]
On my i7 desktop i've got vgg16  inference time 1080ms down to 670ms

**Comment by raver119** [comment]
Inception v3 model imported shows:

> o.d.e.c.PredictWithVGG16 - Last time: 290 ms

don't mind VGG16 word though in logger msg, it's just the same class running inception model.

**Comment by raver119** [comment]
Can someone confirm me please, so we could close this issue?

**Comment by crockpotveggies** [comment]
Confirmed, original VGG-16 feed forward took 1.1 seconds. This is now reduced to 647 milliseconds.

![demo](https://files.gitter.im/deeplearning4j/deeplearning4j/earlyadopters/ntzT/Screen-Shot-2017-02-22-at-5.47.57-PM.png)

**Comment by raver119** [comment]
Thanks for highlighting this issue guys.

I'm closing this issue now, please feel free to open new issue if you'll find something else.
